### PR TITLE
feat(new-post): Add 'is tdd a gimmick?'

### PR DESCRIPTION
**Pull request by jLouzado**

### CODE REVIEW

File: content/blog/2019-10-3-tdd-gimmick/index.md [hunk 1]
```diff
@@ -0,0 +1,196 @@
+---
+title: 'Is TDD just a gimmick?'
+date: '2019-03-24T12:51:00.314Z'
+---
+
+[[toc]]
+
+# Statement
+
+Every day we see new frameworks rise and fall, and a certain amount of skepticism is healthy to keep from jumping on every new thing that comes along.
```
Comment (dlebrero) [inline content/blog/2019-10-3-tdd-gimmick/index.md]: I wouldn't call TDD new. It has been around for at least 18 years in its current form.

File: content/blog/2019-10-3-tdd-gimmick/index.md [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+> Unit tests produced by TDD guarantee code-coverage in a way that tests written post-implementation cannot.
+
+## Maybe it does, but why should I care about code-coverage?
+
+Just to be clear, 100% code-coverage means that if a function is performing `N` operations, each of them are covered by a unit-test. Put another way, if someone were to delete[^del] a line of code atleast one unit-test _should_ fail.
+
+If we have coverage we can have confidence, and with that confidence you get some effects:
+
+- code comprehension will always take linear time even if code-complexity increases exponentially.
```
Comment (dlebrero) [inline content/blog/2019-10-3-tdd-gimmick/index.md]: I personally think this is a bold statement. You still need to understand the "production" code, that if it is complex, it will take a long time to understand. I wouldn't talk about "linear".

What it is very true is that tests will let you understand it better, both because of the documentation that they provide and because, as you mention above, you can "play" with the code (by changing/deleting lines) and see how it affects the test suite.

File: content/blog/2019-10-3-tdd-gimmick/index.md [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+  - the tests will always exhaustively describe the function's behaviour, regardless of how "clever" the implementation is.
```
Comment (dlebrero) [inline content/blog/2019-10-3-tdd-gimmick/index.md]: Nowadays, when I see "exhaustively" I think about property based testing, not in example based testing.

File: content/blog/2019-10-3-tdd-gimmick/index.md [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+First, a quick recap of TDD's rules.
+
+- If you want to write new code (or make changes to an existing unit) first write a failing test.
+  - this is the `RED` phase.
+- Only write as much code as is necessary to produce a failure.
+  - compilation errors count as failures, and
+- Only write as much code as is required to fix that failure, nothing more.
+  - this is the `GREEN` phase
+- Rinse and repeat till all your requirements are implemented.
```
Comment (dlebrero) [inline content/blog/2019-10-3-tdd-gimmick/index.md]: I do refactor as I implement, not at the end of writing all the tests. I do not refactor after each test if I do not consider it necessary, so I would personally move this under the refactor phase and make it obvious that the refactoring is optional.

File: content/blog/2019-10-3-tdd-gimmick/index.md [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+- logical errors (bugs) become much easier to spot and fix, since other people can read your tests and check if all the requirements are being fulfilled.
+- changes to the codebase are guaranteed to not introduce regression bugs.
+
+So confidence (via code-coverage) is the whole game; you will see the above benefits in your codebase to the extent that the requirements are covered, simple as that.
+
+## What does "Produced By TDD" mean?
+
+First, a quick recap of TDD's rules.
+
+- If you want to write new code (or make changes to an existing unit) first write a failing test.
```
Comment (dlebrero) [inline content/blog/2019-10-3-tdd-gimmick/index.md]: I like to make it clear that you are supposed to write one and just one failing test. So I would highlight the "a" or change it for one and only one.

File: content/blog/2019-10-3-tdd-gimmick/index.md [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+  - this is the `RED` phase.
+- Only write as much code as is necessary to produce a failure.
+  - compilation errors count as failures, and
+- Only write as much code as is required to fix that failure, nothing more.
+  - this is the `GREEN` phase
+- Rinse and repeat till all your requirements are implemented.
+- Go back over your implementation and clean-up.
+  - This is the `REFACTOR` phase.
+
```
Comment (dlebrero) [inline content/blog/2019-10-3-tdd-gimmick/index.md:39]: Maybe make it clear on which steps do you run the test suite?

File: content/blog/2019-10-3-tdd-gimmick/index.md [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+}
+```
+
+With that your test will pass, and you can start with the next test!
+
+## But this way of writing tests must be super annoying
+
+Writing code this way is definitely a little unnatural at first, but it makes sure that nothing extraneous ever enters your code base. When people talk about Lean-Testing, and only writing the minimum tests required, this is really what that looks like; no more tests than necessary, no more code than needed.
+
+Over time the intuition for this style will build and your compiler will be the best pair-programming partner you could ever ask for.
```
Comment (dlebrero) [inline content/blog/2019-10-3-tdd-gimmick/index.md]: I think you are not serious about replacing pair programming with TDD, right? 🙈

File: content/blog/2019-10-3-tdd-gimmick/index.md [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+TDD doesn't guarantee that your code is bug-free. It just guarantees that if a line is deleted that a test will fail. Therefore it's still your responsibility to write the correct tests. For example if the requirement was to multiply two numbers, but your tests are checking if the code adds the two together then your code is still wrong.
+
+In such a scenario, TDD helps in two ways:
+
+- it makes such an error easier to spot during code-review
+  - since your `expected` value will clearly be something different than the requirement, someone can flag it as an issue
+- once the error is spotted, it becomes easier to rewrite the bad test or to add tests for missing edge cases.
+
+Since refactoring is easier, this whole process happens much more quickly. And yes, while fixing one issue you're less likely to introduce a a regression even in this new feature.
```
Comment (dlebrero) [inline content/blog/2019-10-3-tdd-gimmick/index.md]: Typo: "introduce a a regression"

File: content/blog/2019-10-3-tdd-gimmick/index.md [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+## This is all horribly complicated! I'm a pretty good programmer, I don't need all this
+
+I felt the same way when first starting out. Even now, when discussing some of the finer points of TDD it can feel a little overly-cautious and pessimistic. The thing is though, if there's more than 2 programmers working on a project it's just not possible to always be available for everyone.
```
Comment (dlebrero) [inline content/blog/2019-10-3-tdd-gimmick/index.md]: It took me a while to digest why the second phrase ("The thing is ...") is here. After re-reading, I see that it has to do with the second phrase in the next paragraph ("In the absence..."). May I suggest to get both statements closer/together?

File: content/blog/2019-10-3-tdd-gimmick/index.md [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+
+Imagine you've written a function, and someone wants to make a change to it. If they knew that the tests were high-quality, they could just figure it out themselves and be fairly confident that they weren't doing anything wrong. In the absence of a test-suite they would (hopefully) come and sit on your head and work things out with you.
+
+In the worst case, they're under a lot of pressure so they make the change that re-introduces some subtle bug that you'd already taken care of initially. The bug might only get discovered months later, at which point maybe the dev who introduced the problematic-change is on leave and the commit message just says "fixed issue" so now you're really screwed.
```
Comment (dlebrero) [inline content/blog/2019-10-3-tdd-gimmick/index.md]: It took me a while to understand this paragraph. I had to read the next to fully understand what you were referring to.

File: content/blog/2019-10-3-tdd-gimmick/index.md [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+
+At it's best, TDD is a way to communicate your decisions without needing to write extensive documentation that no one will ever read (or maintain when they make changes). Good-quality unit tests save such an incredible amount of time, and ultimately that's more time that you have to work in peace.
```
Comment (dlebrero) [inline content/blog/2019-10-3-tdd-gimmick/index.md]: I agree with this halfheartedly. Even if tests document the what, they do not document the why, which some times is what you need to decide if you need to fix the expectation of a test'

File: content/blog/2019-10-3-tdd-gimmick/index.md [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+
+For application development, almost everything is a good candidate for TDD. The one exception for frontend development would be if your function returns a particular DOM structure, there's no point writing a test like `Assert <div>Hello World!</div>` and then having a function that just returns `<div>Hello World!</div>`.
+
+You _could_ do it, but then tomorrow if you need to return `<div>Hello <b>World!</b></div>` should the test really break? I mean, I added an enhancement, the old test should "technically" still be valid.
+
+This doesn't mean that all view-related things cannot be done via TDD though. You might have a rock-solid reducer implemented, but if you forget to actually fire an action when a button is clicked then your app is still buggy. Packages like [react-testing-library](https://github.com/testing-library/react-testing-library) help with that problem.
+
+## Any others?
+
+Yes, if you're writing a new package, and there's two potential APIs that would both work equally fine but they both have different (say) performance implications based on their implementation then that might be a bad use-case for TDD.
```
Comment (dlebrero) [inline content/blog/2019-10-3-tdd-gimmick/index.md]: Shouldn't be this a perfect case for TDD? The tests should be the same, the implementation can vary.

File: content/blog/2019-10-3-tdd-gimmick/index.md [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+
+In the worst case, they're under a lot of pressure so they make the change that re-introduces some subtle bug that you'd already taken care of initially. The bug might only get discovered months later, at which point maybe the dev who introduced the problematic-change is on leave and the commit message just says "fixed issue" so now you're really screwed.
+
+At it's best, TDD is a way to communicate your decisions without needing to write extensive documentation that no one will ever read (or maintain when they make changes). Good-quality unit tests save such an incredible amount of time, and ultimately that's more time that you have to work in peace.
+
+## Are there any cases where one shouldn't use TDD?
+
+For application development, almost everything is a good candidate for TDD. The one exception for frontend development would be if your function returns a particular DOM structure, there's no point writing a test like `Assert <div>Hello World!</div>` and then having a function that just returns `<div>Hello World!</div>`.
```
Comment (dlebrero) [inline content/blog/2019-10-3-tdd-gimmick/index.md]: I would put in this in the realm of tests that suffer from over specification, which comes to the question about what a good test is.

File: content/blog/2019-10-3-tdd-gimmick/index.md [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+
+This doesn't mean that all view-related things cannot be done via TDD though. You might have a rock-solid reducer implemented, but if you forget to actually fire an action when a button is clicked then your app is still buggy. Packages like [react-testing-library](https://github.com/testing-library/react-testing-library) help with that problem.
+
+## Any others?
+
+Yes, if you're writing a new package, and there's two potential APIs that would both work equally fine but they both have different (say) performance implications based on their implementation then that might be a bad use-case for TDD.
+
+In such a case you're really out near the edge of software-development and you're almost in research mode. Once you've sorted out architecture and explored the space sufficiently, that's when TDD might be of more use.
+
+I can't stress this enough though, almost everything related to Application Development can be done via TDD. As much as possible, extract your business logic into pure functions and TDD the heck out of it. You'll thank yourself later.
```
Comment (dlebrero) [inline content/blog/2019-10-3-tdd-gimmick/index.md]: This is very good advice

File: content/blog/2019-10-3-tdd-gimmick/index.md [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+
+Imagine you've written a function, and someone wants to make a change to it. If they knew that the tests were high-quality, they could just figure it out themselves and be fairly confident that they weren't doing anything wrong. In the absence of a test-suite they would (hopefully) come and sit on your head and work things out with you.
+
+In the worst case, they're under a lot of pressure so they make the change that re-introduces some subtle bug that you'd already taken care of initially. The bug might only get discovered months later, at which point maybe the dev who introduced the problematic-change is on leave and the commit message just says "fixed issue" so now you're really screwed.
+
+At it's best, TDD is a way to communicate your decisions without needing to write extensive documentation that no one will ever read (or maintain when they make changes). Good-quality unit tests save such an incredible amount of time, and ultimately that's more time that you have to work in peace.
+
+## Are there any cases where one shouldn't use TDD?
```
Comment (dlebrero) [inline content/blog/2019-10-3-tdd-gimmick/index.md]: As much as I love TDD, there are things that is not worth testing. I think you are right on saying "almost everything", which is also my point of view.

I wrote something regarding being too dogmatic at https://dev.to/danlebrero/the-tragedy-of-100-code-coverage

File: content/blog/2019-10-3-tdd-gimmick/index.md [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+
+## But this way of writing tests must be super annoying
+
+Writing code this way is definitely a little unnatural at first, but it makes sure that nothing extraneous ever enters your code base. When people talk about Lean-Testing, and only writing the minimum tests required, this is really what that looks like; no more tests than necessary, no more code than needed.
+
+Over time the intuition for this style will build and your compiler will be the best pair-programming partner you could ever ask for.
+
+## Cool, but I could just write the function and then write the test afterwards
+
+As far as I've spoken to people, no one disagrees with the need to have some unit tests. That means we all agree that it's possible to write buggy code. That implies that our unit-tests could _also_ have bugs; how would we detect that if we wrote it second?
```
Comment (dlebrero) [inline content/blog/2019-10-3-tdd-gimmick/index.md]: I usually get the complain "if I write the test first, instead of after, I will need to keep updating the tests until I finish the implementation".

**Review by dlebrero — COMMENTED** [review]
Thanks a lot for letting me review it. 

I always try to be extra careful with feedback, but if any comment somehow offend you, please do not take it seriously or personally!

Good luck with your TDD preaching! It is a hard one!

File: content/blog/2019-10-3-tdd-gimmick/index.md (hunk 6 above)
Comment (jLouzado) [reply content/blog/2019-10-3-tdd-gimmick/index.md:39]: is there a prescribed time to run tests? I just set them to run on every save.

File: content/blog/2019-10-3-tdd-gimmick/index.md (hunk 14 above)
Comment (jLouzado) [reply content/blog/2019-10-3-tdd-gimmick/index.md]: 🙏

File: content/blog/2019-10-3-tdd-gimmick/index.md (hunk 11 above)
Comment (jLouzado) [reply content/blog/2019-10-3-tdd-gimmick/index.md]: Usually isn't the why just "because that was the requirement" ? So if the test description is written in terms of the product requirement then it should be easy to see what needs to be changed imo.

File: content/blog/2019-10-3-tdd-gimmick/index.md (hunk 12 above)
Comment (jLouzado) [reply content/blog/2019-10-3-tdd-gimmick/index.md]: made a change, let me know if this is clearer.

File: content/blog/2019-10-3-tdd-gimmick/index.md (hunk 15 above)
Comment (jLouzado) [reply content/blog/2019-10-3-tdd-gimmick/index.md]: great article... included the link as reference. :)

File: content/blog/2019-10-3-tdd-gimmick/index.md (hunk 16 above)
Comment (jLouzado) [reply content/blog/2019-10-3-tdd-gimmick/index.md]: this is a good one, I've tried to add an answer.

File: content/blog/2019-10-3-tdd-gimmick/index.md [hunk 17]
```diff
@@ -0,0 +1,283 @@
...
+As developers I understand that there's a healthy amount of skepticism we all have so let's go through some of the common responses I get when people hear that claim.
+
+## Maybe it does, but why should I care about code-coverage?
+
+Just to be clear, 100% code-coverage means that if a function is performing `N` operations, each of them are covered by a unit-test. Put another way, if someone were to delete[^del] a line of code atleast one unit-test _should_ fail.
+
+If we have coverage we can have confidence, and with that confidence you get some effects:
+
+- effort to comprehend a piece of code will always grow more slowly than the complexity of that code, because:
+  - there's always going to be atleast one example of every potential usage, and
```
Comment (dlebrero) [inline content/blog/2019-10-3-tdd-gimmick/index.md]: typo: atleast --> at<blank>least

Comment (jLouzado) [reply content/blog/2019-10-3-tdd-gimmick/index.md]: is it just me or do those two look the same? 🙈 

I think you mean `at-least` or `at least` instead.

Comment (dlebrero) [reply content/blog/2019-10-3-tdd-gimmick/index.md]: lol. I meant at least. I typed `at<blank>least` and Github decided to clean up my input.

File: content/blog/2019-10-3-tdd-gimmick/index.md [hunk 18]
```diff
@@ -0,0 +1,283 @@
...
+  - you get an opportunity to play around with the code and see how it affects the test-suite.
```
Comment (dlebrero) [inline content/blog/2019-10-3-tdd-gimmick/index.md]: test-suite --> `test suite`

**Comment by dlebrero** [comment]
On every save is perfect, but this is something that people that do not
TDD is not aware of. I think it is important to make it clear for them
when/how often the tests should be run.

**Comment by dlebrero** [comment]
"And why was that the requirement?" is the question that I now struggle
with. But the comment was just for me to raise a pain that I have seen,
which TDD does not solve. Maybe you have a good anwser to record the whys!

**Comment by dlebrero** [comment]
I see what you mean now. Maybe I would change "you're almost in research
mode" for "you're in research mode"

File: content/blog/2019-10-3-tdd-gimmick/index.md (hunk 11 above)
Comment (jLouzado) [reply content/blog/2019-10-3-tdd-gimmick/index.md]: Ref: https://github.com/jLouzado/patchcording.com/pull/11#issuecomment-541662742
